Add real docs

pdeffebach opened the pull request:
This is a major PR (it felt like the last one was, too!)

I am getting us set up with "real" documentation. 

1. Transferred the `README` to `index.md`
1. Added an `api.md` with the `@doc` macros 
1. Fixed all the `jldoctests` so they run

I still need to

1. Write a new `README`
2. Figure out how to get set up so it's a website

Also, I can't get the link to the `api` page to work from the `index` page. If anyone knows how to fix this, please let me know!

cc @bkamins, @nalimilan

bkamins commented:
Great! Let me know when you are done and I will have a look (though @nalimilan is much better than me with documentation).

pdeffebach commented:
Okay I've taken this PR as far as I'm comfortable with for now. 

The changes to the docs are minimal. But lots of examples have been changed to no longer rely on random number generation. 

Next step is hosting. If this were my own package I would fumble through the hosting configuration on my own. But given this is owned by JuliaData I would like someone, probably @quinnj  or @nalimilan to help take this over the finish line. in particular, the part about adding an ssh key (which I have, I think) and making sure a website actually *exists* confuses me. 

Thank you!

quinnj commented:
Usually it's not too bad to setup; I just follow the Documenter.jl guide for hosting; in travis, you put the DOCUMENTER_KEY as an environment variable (I can do this if you dm me the key), then make sure your travis job has a docs build job.

pdeffebach commented:
@quinnj or @nalimilan 

I've worked through the documenter docs on hosting, and i need to add a ssh key for the repository. I don't have ownership permissions to do that for DataFramesMeta (can't click on settings at the repo page on github). I need one of you to work through [this section](https://juliadocs.github.io/Documenter.jl/stable/man/hosting/#travis-ssh) to set up the SSH deploy keys.

pdeffebach commented:
Okay I think this is ready for another review. I have 

* Added the ssh key to the Travis settings
* Added the ssh key to the DataFramesMeta settings
* Added the deployment ssh command to the tagbot yml
* Deleted an extraneous appveyer badge
* Added badges for the documentation
* Written a little more on the README

I think that if we merge this the docs should deploy correctly. 

Please let me know!

pdeffebach commented on docs/src/api/api.md:
````diff
@@ -0,0 +1,34 @@
+# API
+
+```@docs
+@transform(x, args...)
````
Thanks! didn't know about `@autodocs`. It's probably good to document deprecated things anyways so people can see them in a search.

pdeffebach commented on src/DataFramesMeta.jl:
```diff
@@ -596,19 +599,19 @@ function combine_helper(x, args...; deprecation_warning = false)
 
 ### Arguments
 
 * `g` : a GroupedDataFrame
```
thanks. added.

pdeffebach commented on src/DataFramesMeta.jl:
```diff
@@ -618,34 +621,13 @@ julia> g = groupby(d, :x);
-│ 2   │ 2 │ 84   │
-│ 3   │ 3 │ 27   │
-
-julia> @combine(g, x2 = 2 * :x, nsum = sum(:n))
```
Yes, it shows the "spreading". re-added.

pdeffebach commented:
Tests pass!

bkamins commented on docs/src/index.md:
```diff
@@ -0,0 +1,457 @@
+and [LINQ](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/)
+in C#. 
+
+In addition, DataFramesMeta provides 
```
```suggestion
In addition, DataFramesMeta.jl provides 
```

bkamins replied:
same in other places

pdeffebach replied:
Trying to only use the `.jl` in introductions, and just the package name after.

bkamins replied:
OK - we can go this way if you prefer

bkamins commented on docs/src/index.md:
```diff
@@ -0,0 +1,457 @@
+# Introduction
+
+Metaprogramming tools for DataFrames.jl objects.
+These macros improve performance and provide more convenient syntax.
```
maybe add 1-2 examples of DataFrames.jl code and equivalent DataFramesMeta.jl code and say that they show this convenience.

Also maybe comment in what cases the macros improve performance somewhere (probably not here)

pdeffebach replied:
Added an example of the convenience of DataFramesMeta. 

I dont think any of them improve performance over DataFrames. I will defer writing this section and delete the claim at the top of the document.

bkamins commented on docs/src/index.md:
```diff
@@ -0,0 +1,457 @@
+
+## `@eachrow`
+
+Act on each row of a data frame. Includes support for control flow and `begin end` blocks. Since the "environment" induced by `@eachrow df` is implicitly a single row of `df`, one uses regular operators and comparisons instead of their elementwise counterparts as in `@with`. Does not change the input data frame argument.
```
add line breaks

bkamins commented on docs/src/index.md:
```diff
@@ -0,0 +1,457 @@
+
+`@eachrow` also supports special syntax for allocating new columns to make
+`@eachrow` more useful for data transformations. The syntax `@newcol
+x::Array{Int}` allocates a new column `:x` with an `Array` container with eltype
```
This is the first time I notice it. Why `Array{Int}` not `Vector{Int}`? Also can you allocate other types than vectors from Julia Base (e.g. `PooledVector`)?

pdeffebach replied:
You can allocate any type that would allow this pattern 

```
T(undef, N)
```

So it would probably be pretty hard to construct other types of arrays, I think. 

This is something I would really like to fix but haven't thought of a way to do so without type instabilities.

bkamins replied:
OK - clear. We can leave it for later. Now we document that `Array` is only allowed, so we have flexibility for the future.

pdeffebach replied:
I have changed all examples to array.

bkamins commented on docs/src/index.md:
```diff
@@ -0,0 +1,457 @@
+avoids filling up the limited macro name space. The main downside is
+that more magic happens under the hood.
+
+This method is extensible. Here is a comparison of the macro and
```
is this comparison or an example how one could write ones own extension?

pdeffebach replied:
I don't think I want to expose this API formally yet. Its untested, and I don't know if there is a need to someone to write their own extension. I think we should cut this.

bkamins replied:
OK - then I would cut this.

bkamins commented on docs/src/index.md:
```diff
@@ -0,0 +1,457 @@
+Again, this is experimental. Based on feedback, we may decide to only
+use `@linq` or only support the set of linq-like macros.
+
+Alternatively you can use Lazy.jl `@>` macro like this:
```
do we really need to show Lazy.jl here? What is the value of this?

pdeffebach replied:
I agree. I'm fine to delete.

bkamins commented on docs/src/index.md:
```diff
@@ -0,0 +1,457 @@
+select number of functions into the namespace by only importing `@>` e.g. 
+`using Lazy: @>` instead of `using Lazy`.
+
+Another alternative is Pipe.jl which exports the `@pipe` macro for piping. 
```
same with Pipe.jl. I like and use it with DataFrames.jl but what is the benefit here? Do we recommend it?

pdeffebach replied:
Actually, I think I will keep the piping sections. `@>` and `@pipe` have different behaviors, so it makes sense for users to choose what they are most comfortable with, for now. 

Eventually I would like DataFramesMeta to provide one really good macro and not recommend others, but that hasn't happened yet.

bkamins commented:
Looks good. Thank you. I have left some minor comments.

pdeffebach commented:
Thanks!

I deleted the example about extending `@linq` but am keeping other piping. `@linq` isn't perfect so users might want to know about other options. 

Ready to merge if people approve!

nalimilan commented on docs/make.jl:
```diff
@@ -3,6 +3,10 @@ using Documenter, DataFramesMeta
 	sitename = "DataFramesMeta Documentation",
+	format = Documenter.HTML(
+		canonical = "https://juliadata.github.io/DataFramesMeta.jl/stable/",
+		assets = ["assets/favicon.ico"]
```
I think you need to add the icon to docs/src/assets/ (you can copy it from DataFrames). Or just drop this line.

pdeffebach commented:
Thanks for your feedback, everyone! I am going to merge!
